faster pairwise summation

John-Colvin opened the pull request:
Using an iterative approach instead of recursive. When I originally wrote this I recorded an order-of-magnitude speedup on arrays and approx. 2x on more complicated ranges.

andralex commented on std/algorithm/iteration.d:
```diff
@@ -4035,16 +4038,98 @@ if (isInputRange!R && !isInfinite!R && is(typeof(seed = seed + r.front)))
+                data = data[16 .. $];
+            }
+            else store[idx] = sumPairwiseN!(16, false, F)(data);
+            foreach (_; 0 .. cast(uint)bsf(k+1))
```
This should best be a while loop checking the limit of `idx` so there's only one iteration variable.

John-Colvin replied:
done, the others as well

andralex commented on std/algorithm/iteration.d:
```diff
@@ -4035,16 +4038,98 @@ if (isInputRange!R && !isInfinite!R && is(typeof(seed = seed + r.front)))
+        foreach (el; data)
+        {
+            store[idx] = el;
+            foreach (_; 0 .. cast(uint)bsf(i+1))
```
same here

andralex commented on std/algorithm/iteration.d:
```diff
@@ -4035,16 +4038,98 @@ if (isInputRange!R && !isInfinite!R && is(typeof(seed = seed + r.front)))
+        while (!data.empty)
+        {
+            store[idx] = sumPairwiseN!(16, true, F)(data);
+            foreach (_; 0 .. cast(uint)bsf(k+1))
```
and here :)

John-Colvin commented:
@andralex Thanks for the comments, I'll fix them soon. Are you interested in more things like this? I do love writing fast implementations of simple algorithms, but e.g. if you compare this implementation to the old one, it's a lot harder to understand / maintain.

andralex commented:
@John-Colvin depends on the gains. At some point we do need to add some benchmarks to Phobos, and let's defer performance work to after that.

In this case I'd say eliminating the unused iteration variable may arguably lead to code that's easier to understand and maintain, too.

John-Colvin commented:
@andralex I don’t mean your suggestions make it harder to maintain/understand (quite the opposite actually), I was comparing to the old code, which was very simple.

If we had a framework for benchmarking then each performance improvement pull could be required to add an attempt at best/worst/mediocre case benchmarks for the new implementation. It’s pretty boring writing benchmarks without being allowed to optimise the code, so it might be good to bundle the work together.

> On 10 Mar 2016, at 15:07, Andrei Alexandrescu <notifications@github.com> wrote:
> 
> @John-Colvin <https://github.com/John-Colvin> depends on the gains. At some point we do need to add some benchmarks to Phobos, and let's defer performance work to after that.
> 
> In this case I'd say eliminating the unused iteration variable may arguably lead to code that's easier to understand and maintain, too.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/D-Programming-Language/phobos/pull/4069#issuecomment-194892209>.
>

John-Colvin commented:
Ewwww, found a nasty bug in `std.range.Take`. Fixed, tested, added more testing for pairwise sum (which was how I found the bug, completely by chance), changed to use one less loop-variable as per @andralex's recommendation and factored the whole store-collapse in to a local function.

John-Colvin commented:
The std.range.Take bug was introduced in https://github.com/D-Programming-Language/phobos/pull/2158, which is a great example of why unittests should actually test something, not just do something to check it compiles.

John-Colvin commented on std/algorithm/iteration.d:
```diff
@@ -4035,16 +4038,97 @@ if (isInputRange!R && !isInfinite!R && is(typeof(seed = seed + r.front)))
+    F[64] store = void;
+    size_t idx = 0;
+
+    auto collapseStore(T)(T k)
```
@andralex I had to make this a template because `size_t` is no good as a generic index/length type for ranges (e.g. sdt.range.iota of ulong on 32bit). I now don't know what to do with e.g. line 4087 which uses size_t as counter. Should I use `ulong`? What is the maximum allowed length for a range?

andralex replied:
My thinking in this matter is to avoid implementation aggravation. If someone needs ranges longer than 4B, they need to switch to a 64 bit platform. I think dealing with lengths that are non-`size_t` is just wrong.

jmdavis replied:
Agreed. It overcomplicates life to support lengths of any type other than `size_t`.

John-Colvin replied:
Ok, that's fine, I just needed a decision. I'll sort this out now.

John-Colvin replied:
on second thoughts, what does that mean for `iota(5L)` or `iota(long.max) on 64 bit? What type should the length be?

9il replied:
> on second thoughts, what does that mean for iota(5L) or iota(long.max) on 32 bit? What type should the length be?

Hide this kind of thoughts! They are dangerous XDDDD
Seriously I don't know. But I would like summation algorithms work only with `size_t`.

quickfur replied:
IMO, if we go with enforcing `length` to be `size_t`, then `iota(long.max)` should simply have no `length`.  It's a small price to pay for a simple and consistent API.  I've considered generalizing `length` to allow arbitrary arithmetical types, but one quickly gets lost in a forest of how to update existing code that assumes `size_t` and how to write generic code that can work with `length` of any type, etc.. The ROI doesn't justify it.

9il replied:
> IMO, if we go with enforcing length to be size_t, then iota(long.max) should simply have no length

+1

schveiguy replied:
This is pretty much already addressed. Somehow got stalled: #4013

andralex replied:
Iota with longs should probably not offer length.

John-Colvin replied:
@andralex On x86 or on x64 as well?
On 24 Mar 2016 8:04 pm, "Andrei Alexandrescu" <notifications@github.com>
wrote:

> In std/algorithm/iteration.d
> <https://github.com/D-Programming-Language/phobos/pull/4069#discussion_r57379092>
> :
>
> >  {
> > -    static assert (isFloatingPoint!Result);
> > -    switch (r.length)
> > +    import core.bitop : bsf;
> > +    // Works for r with at least length < 2^^(64 + log2(16)), in keeping with the use of size_t
> > +    // elsewhere in std.algorithm and std.range on 64 bit platforms. The 16 in log2(16) comes
> > +    // from the manual unrolling in sumPairWise16
> > +    F[64] store = void;
> > +    size_t idx = 0;
> > +
> > +    auto collapseStore(T)(T k)
>
> Iota with longs should probably not offer length.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/D-Programming-Language/phobos/pull/4069/files/ef9f8338fe123c1514a61c5b4b50fbfb072632e4#r57379092>
>

jmdavis replied:
#4013 takes the approach of asserting that the length of the range will fit in `size_t`, so `long` and `ulong` will work on 32-bit systems - but only if the resultant range is not too long. So, most common cases will work just fine, but if you try and do something like `iota(long.min, long.max)`, then the assertion will fail. The downside, of course, is that when the problem does occur, it's only found at runtime, but in the vast majority of cases, you can continue to use `long` and `ulong` with `iota` and have `length` as we've had.

9il commented:
I am working on summation module... could we hold this PR?

9il commented:
@klickverbot Ok

greenify commented:
Please rebase as it now has merge conflicts.

John-Colvin commented:
rebased

monarchdodra commented on std/range/package.d:
```diff
@@ -1841,9 +1841,9 @@ if (isInputRange!(Unqual!Range) &&
+            assert(j <= length, "Attempting to slice past the end of a "
                 ~ Take.stringof);
-            return source[i .. j - i];
+            return source[i .. j];
```
Oops. Not sure what I was thinking when I wrote this. Kind of sad it took 2 years to catch it :/

9il commented:
`mir.las.sum` was released. May be we can just put the link to mir in the docs :-) ? I expect that `mir` version would be significantly faster, especially with optimization for native CPU by LDC/GDC.

quickfur commented:
Should we just close this then?  Or does somebody want to port the mir version to Phobos?

greenify commented:
> Should we just close this then? Or does somebody want to port the mir version to Phobos?

[AFAIK](https://github.com/DlangScience/mir/issues/28) will @9il merge mir periodically to Phobos.

9il commented:
> AFAIK will @9il merge mir periodically to Phobos.

Nitpick: without official review process (see wiki) I can not add new modules.

greenify commented:
@John-Colvin could you please separate your fix for `Take` into another PR?

JackStouffer commented:
I argue for this to be merged. Just because something in the future might be faster doesn't mean that improving what we have not shouldn't happen. This is a performance optimization of some pretty common code, it should be merged posthaste.

9il commented:
> @John-Colvin could you please separate your fix for Take into another PR?

This is OK to have just a commit without separate PR. In the same time we need a filled issue to show this fix in the history.

9il commented:
If someone would like to have this in Phobos feel free to merge.

Niptick: this algorithms allows to remove Kahan summation, so it should be removed.

9il commented:
Oh, looks like a may be wrong about performance for current state of mir :-/

greenify commented:
> This is OK to have just a commit without separate PR. In the same time we need a filled issue to show this fix in the history.

OK, sorry to be an annoying complainer ...

9il commented:
### Test results
The PR code is better for DMD and Mir `0.11.0-beta2` is better for LDC. The difference for LDC is small if bounds check are turned off and array is large. Mir is 2x+ times faster when array length = 15.

So, the PR LGTM. Do not forget to remove Kahan summation.

| Compiler | BoundsCheck | Array Length | Bench Count | Algorithm | Time |
|-----|-----|-----|-----|-----|-----|
| DMD | safe | 10000 | 10000 | std | 466 ms |
| DMD | safe | 10000 | 10000 | PR | 94 ms |
| DMD | safe | 10000 | 10000 | Mir | 128 ms |
| DMD | safe | 15 | 100000 | std | 69 ms |
| DMD | safe | 15 | 100000 | PR | 73 ms |
| DMD | safe | 15 | 100000 | Mir | 65 ms |
| DMD | off | 10000 | 10000 | std | 407 ms |
| DMD | off | 10000 | 10000 | PR | 43 ms |
| DMD | off | 10000 | 10000 | Mir | 96 ms |
| DMD | off | 15 | 100000 | std | 55 ms |
| DMD | off | 15 | 100000 | PR | 61 ms |
| DMD | off | 15 | 100000 | Mir | 63 ms |
| LDC | safe | 10000 | 10000 | std | 306 ms |
| LDC | safe | 10000 | 10000 | PR | 62 ms |
| LDC | safe | 10000 | 10000 | Mir | 38 ms |
| LDC | safe | 15 | 100000 | std | 38 ms |
| LDC | safe | 15 | 100000 | PR | 59 ms |
| LDC | safe | 15 | 100000 | Mir | 26 ms |
| LDC | off | 10000 | 10000 | std | 307 ms |
| LDC | off | 10000 | 10000 | PR | 39 ms |
| LDC | off | 10000 | 10000 | Mir | 37 ms |
| LDC | off | 15 | 100000 | std | 40 ms |
| LDC | off | 15 | 100000 | PR | 56 ms |
| LDC | off | 15 | 100000 | Mir | 22 ms |

9il commented:
Benchmark code http://dpaste.dzfl.pl/adb07eea2db4

DmitryOlshansky commented:
Anyhow it would be good to merge this while waiting on more of MIR additions to Phobos

wilzbach commented:
Ping - so any other objections except that khan summation should be removed? I also vote for merging :)

9il commented:
> Ping - so any other objections except that khan summation should be removed? I also vote for merging :)

I have not. Vote too

DmitryOlshansky commented:
Auto-merge toggled on

9il commented:
> Auto-merge toggled on

Why Kahan was not removed?

DmitryOlshansky commented:
> Why Kahan was not removed?

Was that a requirement?

9il commented:
> Was that a requirement?

New pairwise summation can work with input ranges, so we don't need Kahan summation algorithms, which was selected of range has hot random access

wilzbach commented:
Should I sent a follow-up PR to remove Kahan?

DmitryOlshansky commented:
Yeah, a follow-up makes sense

wilzbach commented:
> New pairwise summation can work with input ranges, so we don't need Kahan summation algorithms, which was selected of range has hot random access

Just had a closer look at it and didn't see the random access

```
 static assert (isFloatingPoint!Result && isMutable!Result);
    Result c = 0;
    for (; !r.empty; r.popFront())
    {
        auto y = r.front - c;
        auto t = result + y;
        c = (t - result) - y;
        result = t;
    }
    return result;
```

and

> $(LI If $(D ElementType!R) is a floating-point type and $(D R) is a
finite input range (but not a random-access range with slicing), then
$(D sum) uses the $(WEB en.wikipedia.org/wiki/Kahan_summation,
Kahan summation) algorithm.)

John-Colvin commented:
For those interested in speed, there are two things I discovered:

LDC doesn't inline core.bitop.bsf
manually unrolling to 32 is even faster than 16

I found that copy-pasting the bsf definition in to the benchmark module and replacing sumPairwise16 with sumPairwise32 brought this implementation to speed-parity with mir.

9il commented:
> I found that copy-pasting the bsf definition in to the benchmark module

We need to make `bsf` and others templates

John-Colvin commented:
that or fix cross-module inlining

9il commented:
> ...  or fix cross-module inlining

This would not work for Phobos. Even if you place `bsf` in module with summation `bsf` would be called from Phobos.

John-Colvin commented:
Hmm, GDC and DMD manage to inline it ok. I guess they're magic intrinsics.

Any idea what's going on with the definition in dmd's druntime? Looks
seriously weird to me. On x86_64 size_t is ulong, so there's a separate
declaration and definition in the same module, in D? Then the definition
does a no-op cast and calls itself? Whut? I'm guessing the way intrinsics
work in DMD breaks language rules and is confusing me.
On 12 Apr 2016 5:36 pm, "Ilya Yaroshenko" <notifications@github.com> wrote:

> ... or fix cross-module inlining
>
> This would not work for Phobos. Even if you place bsf in module with
> summation bsf would be called from Phobos.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/D-Programming-Language/phobos/pull/4069#issuecomment-208996521>
>
